euphoria-flink/src/main/java/cz/seznam/euphoria/flink/accumulators/FlinkNativeAccumulators.java

```diff
@@ -0,0 +1,175 @@
+        throw new RuntimeException("Exception during accumulator initialization: " + clz, e);
+      }
+    }
+
+    if (!clz.isAssignableFrom(acc.getClass())) {
+      throw new IllegalStateException(
+              "Accumulator named '" + name + "' is type of " + acc.getClass().getSimpleName());
+    }
+
+    return (ACC) acc;
+  }
+
+  public static Factory getFactory() {
+    return Factory.get();
+  }
+
+  // ------------------------------
+
+  public static class Factory implements FlinkAccumulatorFactory {
+
+    private static final Factory INSTANCE = new Factory();
+
+    private Factory() {}
+
+    @Override
+    public AccumulatorProvider create(Settings settings, RuntimeContext context) {
+      return new FlinkNativeAccumulators(context);
+    }
+
+    public static Factory get() {
+      return INSTANCE;
+    }
+  }
+
+  // ------------------------------
+
+  // Each class implements both Flink Accumulator and Euphoria Accumulator
+  // so it can be used in both contexts without creating multiple instances.
+
+  public static class FlinkCounter extends LongCounter implements Counter {
+
+    @Override
+    public void increment(long value) {
+      super.add(value);
+    }
+
+    @Override
+    public void increment() {
+      super.add(1L);
+    }
+
+    // It is necessary to override clone method to instantiate this
+    // extended class and not the parent type.
+    @Override
+    @SuppressWarnings("CloneDoesntCallSuperClone")
+    public LongCounter clone() {
+      FlinkCounter result = new FlinkCounter();
+      result.merge(this);
+      return result;
+    }
+  }
+
+  public static class FlinkHistogram extends org.apache.flink.api.common.accumulators.Histogram
+          implements Histogram {
+
+    @Override
+    public void add(long value) {
+      super.add(Math.toIntExact(value));
+    }
+
+    @Override
+    public void add(long value, long times) {
+      for (int i = 0; i < times; i++) {
+        this.add(value);
+      }
+    }
+
+    // It is necessary to override clone method to instantiate this
+    // extended class and not the parent type.
+    @Override
+    @SuppressWarnings("CloneDoesntCallSuperClone")
+    public org.apache.flink.api.common.accumulators.Accumulator<Integer, TreeMap<Integer, Integer>> clone() {
+      FlinkHistogram result = new FlinkHistogram();
+      result.merge(this);
+      return result;
+    }
+  }
+
+  public static class FlinkTimer extends org.apache.flink.api.common.accumulators.Histogram
+          implements Timer {
+
+    @Override
+    public void add(Duration duration) {
+      super.add(Math.toIntExact(duration.toMillis()));
+    }
+
+    // It is necessary to override clone method to instantiate this
+    // extended class and not the parent type.
+    @Override
+    @SuppressWarnings("CloneDoesntCallSuperClone")
+    public org.apache.flink.api.common.accumulators.Accumulator<Integer, TreeMap<Integer, Integer>> clone() {
+      FlinkTimer result = new FlinkTimer();
```

Comment:
I am not sure about this. The clone method override is enforced by the Flink's superclass `Histogram`. Creating a copy constructor wouldn't solve anything in this case. But I agree this is a kind of ugly hack - at least properly described in code comment.